### PR TITLE
TransferFiles To Custom Directory

### LINKED ISSUE #38865: [FR] Make destination folder for a files:transfer-ownership configurable
### Actual behaviour
We have to migrate about 500 accounts, where we will merging the various departments to only one company. All this users will get a new username, where we will move the full account include file locations from username A to username B.

In the current situation, when you migrate a user via files:transfer-ownership, all files will end up in a subfolder like ; `transferred from old-username on 2020618..`

### Steps to reproduce
1. `occ files:transfer-ownership old-username new-username`

### Expected behaviour
Would be nice if the destination path becomes configurable, so that a transfer can end up elsewhere. For example in a subfolder with custom name or directly in the root for a new account.

### PR DESCRIPTION
**Pull request by T0mWz**
## Description
When you migrate a user via files:transfer-ownership, all files will end up in a subfolder , for example ; `transferred from old-username on 2020618`.
Would be nice if the destination path becomes configurable, so that a transfer can end up elsewhere. For example in a subfolder with custom name or directly in the root when you move to a new account.

## Related Issue
- Fixes https://github.com/owncloud/core/issues/38865

## How Has This Been Tested?

- test environment: Acceptance environment
- test case 1: without arguments
- test case 2: with empty argument
- test case 3: with an argument 

## Screenshots (if appropriate):
```
root@vm-web-acc-01:/var/www/owncloud # sudo -u apache /usr/bin/php72 /var/www/owncloud/occ files:transfer-ownership --destination-path='' old-user new-user
Analysing files of wezep001@surf.nl ...
   13 [============================]
Collecting all share information for files and folder of old-user ...
    0 [>---------------------------]
Transferring files to new-user/files/ ...
Restoring shares ...
    0 [>---------------------------]

```
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [ ] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

### CODE REVIEW

**Comment by phil-davis** [comment]
Note: this can be tested by adding scenarios to https://github.com/owncloud/core/blob/master/tests/acceptance/features/cliMain/transfer-ownership.feature
which will require adapting test code in https://github.com/owncloud/core/blob/master/tests/acceptance/features/bootstrap/TransferOwnershipContext.php etc.

File: apps/files/lib/Command/TransferOwnership.php [hunk 1]
```diff
@@ -126,6 +129,12 @@ protected function configure() {
 				InputOption::VALUE_REQUIRED,
 				'selectively provide the path to transfer. For example --path="folder_name"'
 			)
+			->addOption(
+				'destination-path',
+				null,
+				InputOption::VALUE_REQUIRED,
+				'selectively provide the path where the data will be transferred to. For example --destination-path="folder_name"'
```
Comment (mmattel) [inline apps/files/lib/Command/TransferOwnership.php:136]:
```suggestion
				'Selectively provide the path where the data will be transferred to. For example --destination-path="folder_name"'
```

File: apps/files/lib/Command/TransferOwnership.php [hunk 2]
```diff
@@ -151,6 +160,16 @@ protected function execute(InputInterface $input, OutputInterface $output) {
 		$this->inputPath = $input->getOption('path');
 		$this->inputPath = \ltrim($this->inputPath, '/');
 
+		$this->destinationPath = $input->getOption('destination-path');
+		if ($this->destinationPath === null) {
+			\OC::$server->getLogger()->debug('When destinationPath not defined we will use the default');
```
Comment (mmattel) [inline apps/files/lib/Command/TransferOwnership.php:165]:
```suggestion
			\OC::$server->getLogger()->debug('The destinationPath is not defined, using the default one');
```
based on the query, it should result in a statement not a question.

**Comment by mmattel** [comment]
Docs relevant, pls file a issue in docs when close to merge.
@EParzefall fyi

Regarding the definable path, pls write which restrictions or prerequisite(s) apply (eg, anywhere in the filesystem, access rights ect)

File: apps/files/lib/Command/TransferOwnership.php (hunk 2 above)
Comment (jvillafanez) [reply apps/files/lib/Command/TransferOwnership.php:165]: I don't think we need this log. Same with the log below.

File: apps/files/lib/Command/TransferOwnership.php [hunk 3]
```diff
@@ -151,6 +160,16 @@ protected function execute(InputInterface $input, OutputInterface $output) {
...
+			$date = \date('Ymd_his');
+			$this->destinationPath = "transferred from $this->sourceUser on $date";
+		} else {
+			\OC::$server->getLogger()->debug('Custom destinationPath is defined, will using: ' . $this->destinationPath);
+			$this->destinationPath = \ltrim($this->destinationPath, '/');
```
Comment (jvillafanez) [inline apps/files/lib/Command/TransferOwnership.php:170]: `trim($this->destinationPath, '/')` seems safer.

We can also reorder this a bit:
```
$this->destinationPath = "transferred from $this->sourceUser on $date";
$destinationPathOpt = $input->getOption('destination-path');
if ($destinationPathOpt !== null) {
  $this->destinationPath = \trim($destinationPathOpt, '/');
}
```

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>webUISharingIntGroup1-chrome-mariadb10.2-php7.4</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>apiWebdavOperations-mariadb10.2-php7.4</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>apiWebdavLocks-mariadb10.2-php7.4</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>apiWebdavMove2-mariadb10.2-php7.4</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>apiWebdavProperties2-mariadb10.2-php7.4</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>webUISettingsMenu-chrome-mariadb10.2-php7.4</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>apiSharingNotificationsToShares-mariadb10.2-php7.4</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>webUISharingAutocomp2-chrome-mariadb10.2-php7.4</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>cliBackground-mariadb10.2-php7.4</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>webUISharingAcceptSh-chrome-mariadb10.2-php7.4</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>apiFederationToShares2-git-mariadb10.2-php7.4</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>apiWebdavPreviews-mariadb10.2-php7.4</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>apiWebdavLocks2-mariadb10.2-php7.4</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>cliMain-mariadb10.2-php7.4</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>webUICreateDelete-chrome-mariadb10.2-php7.4</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>apiFederationToRoot1-latest-mariadb10.2-php7.4</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>apiWebdavLocks3-mariadb10.2-php7.4</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>webUIComments-chrome-mariadb10.2-php7.4</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>apiWebdavProperties1-mariadb10.2-php7.4</strong> failed. The build is cancelled...</summary>

**Comment by mmattel** [comment]
I have rebased and squashed the commits. Rebasing was necessary due to: https://github.com/owncloud/core/pull/38973

**Review by mmattel — APPROVED** [review]
LGTM from a text pov 👍

**Comment by mmattel** [comment]
To have a proper documentation preparation, pls tell the restrictions and axamples of the paths used.

**Comment by mmattel** [comment]
Please create a **changelog** before merging !

File: apps/files/lib/Command/TransferOwnership.php [hunk 4]
```diff
@@ -164,9 +180,8 @@ protected function execute(InputInterface $input, OutputInterface $output) {
 			return 2;
 		}
 
-		// use a date format compatible across client OS
```
Comment (JanAckermann) [inline apps/files/lib/Command/TransferOwnership.php:167]: Maybe we can make the destinationPath optional, so we can keep the old behaviour with less required ?

**Review by AlexAndBear — CHANGES_REQUESTED** [review]

**Review by mrow4a — CHANGES_REQUESTED** [review]
needs rebase , generally might evolve to quite complicated change but probably doable

**Comment by mmattel** [comment]
@T0mWz do you have any update on this?

**Comment by DeepDiver1975** [comment]
cleaning up - I think this is no longer of anybody's interest

**Pull request closed without merging**